### PR TITLE
[develop] Add lookup verifier index

### PR DESCRIPTION
**Pull request by dannywillems**
TODO:
- Add corresponding JS converters:
  - [x] WasmLookupVerifierIndex
  - [x] WasmLookupSelectors
  - [x] WasmLookInfo
  - [x] WasmLookupPatterns
  - [x] WasmLookupFeatures

### CODE REVIEW

**Comment by rbonichon** [comment]
I think this was already done (albeit a little bit differently) in #38. We need to reconcile both versions.

**Review by mrmr1993 — APPROVED** [review]

File: kimchi/js/bindings.js [hunk 1]
```diff
@@ -1878,7 +2015,15 @@ var caml_plonk_verifier_index_of_rust = function (x, affine_class) {
   var shifts = caml_plonk_verification_shifts_of_rust(x.shifts);
   // TODO: Handle linearization correctly!
   // var linearization = linearization_of_rust(x.linearization, affine_class);
-  var lookup_index = None;
+  // TODO: converter
```
Comment (dannywillems) [inline kimchi/js/bindings.js:2018]: can be removed

File: kimchi/js/bindings.js [hunk 2]
```diff
@@ -1889,43 +2034,207 @@ var caml_plonk_verifier_index_of_rust = function (x, affine_class) {
...
+  return caml_array_to_rust_vector(
+    wasm_lookup_table,
+    rust_affine_of_caml_affine,
+    mk_affine
+  );
+};
+
+// Provides:: caml_table_ids_to_rust
+// Requires: plonk_wasm
+var caml_table_ids_to_rust = function (wasm_table_ids) {};
```
Comment (dannywillems) [inline kimchi/js/bindings.js:2081]: ?

Comment (rbonichon) [reply kimchi/js/bindings.js:2081]: Useless. Needs to be removed.

Comment (rbonichon) [reply kimchi/js/bindings.js]: Fixed.

File: kimchi/js/bindings.js (hunk 1 above)
Comment (rbonichon) [reply kimchi/js/bindings.js]: Fixed

File: kimchi/js/bindings.js [hunk 3]
```diff
@@ -1866,6 +1857,152 @@ function linearization_of_rust(linearization, affine_class) {
...
+var caml_opt_of_rust = function (value, value_of_rust) {
+  if (value == undefined) {
+    return None;
+  } else {
+    return [0, value_of_rust(value)];
+  }
+};
+
+// Provides: caml_bool_of_rust
+var caml_bool_of_rust = function (bool) {
```
Comment (dannywillems) [inline kimchi/js/bindings.js:1889]: Not sure it is correct. I would say simply `bool`: https://rustwasm.github.io/wasm-bindgen/reference/types/bool.html.

File: kimchi/js/bindings.js [hunk 4]
```diff
@@ -1866,8 +1857,154 @@ function linearization_of_rust(linearization, affine_class) {
...
+  return v === None;
+};
+
+// Provides: caml_map_of_rust_vector
+// Requires: js_class_vector_of_rust_vector
+var caml_map_of_rust_vector = function (v, klass, converter_to_rust) {
+  var a = js_class_vector_of_rust_vector(v, klass);
+  var res = [0];
+  for (var i = 0; i < a.length; ++i) {
+    // TODO Check this. Could be off by 1
```
Comment (dannywillems) [inline kimchi/js/bindings.js:1872]: Let's also remove this.

**Comment by dannywillems** [comment]
Before merging, I would suggest squashing bindings.js changes into only one. In particular, it will remove the ugly `WIP` in the history.

File: kimchi/js/bindings.js [hunk 5]
```diff
@@ -1889,43 +2031,203 @@ var caml_plonk_verifier_index_of_rust = function (x, affine_class) {
...
   ];
 };
+
+// Provides: caml_opt_to_rust
+// Requires: caml_is_none, None
+var caml_opt_to_rust = function (caml_optional_value, to_rust) {
+  if (caml_is_none(caml_optional_value)) {
+    return undefined;
+  } else {
+    return to_rust(caml_optional_value);
```
Comment (dannywillems) [inline kimchi/js/bindings.js:2044]: I would add a single sentence as a documentation to describe if the function `to_rust` should expect `x` in the case of `Some x`. A some value is represented as `[0, vx]` where `vx` is the actual value.
I think it should be `caml_optional_value[1]`.

File: kimchi/js/bindings.js [hunk 6]
```diff
@@ -1866,8 +1857,154 @@ function linearization_of_rust(linearization, affine_class) {
...
+    res.push(converter_to_rust_rust(a[i]));
+  }
+  return res;
+};
+
+// Provides: caml_opt_of_rust
+// Requires: None
+var caml_opt_of_rust = function (value, value_of_rust) {
+  if (value == undefined) {
+    return None;
```
Comment (dannywillems) [inline kimchi/js/bindings.js:1882]: As a comment, comes from:
```
"immediate" (e.g. true, false, None, ()) | number (32bit int)
```
Source: https://github.com/ocsigen/js_of_ocaml/

File: kimchi/js/bindings.js [hunk 7]
```diff
@@ -1866,8 +1857,154 @@ function linearization_of_rust(linearization, affine_class) {
...
+    // TODO Check this. Could be off by 1
+    res.push(converter_to_rust_rust(a[i]));
+  }
+  return res;
+};
+
+// Provides: caml_opt_of_rust
+// Requires: None
+var caml_opt_of_rust = function (value, value_of_rust) {
+  if (value == undefined) {
```
Comment (dannywillems) [inline kimchi/js/bindings.js:1881]: maybe `===`? It is semantically the same for `undefined`, I think.

File: kimchi/js/bindings.js (hunk 5 above)
Comment (dannywillems) [reply kimchi/js/bindings.js:2044]: Source: https://v2.ocaml.org/manual/intfc.html#ss:c-concrete-datatypes

File: kimchi/js/bindings.js (hunk 7 above)
Comment (dannywillems) [reply kimchi/js/bindings.js]: Fixed in [87b72e7](https://github.com/o1-labs/snarkyjs-bindings/pull/114/commits/87b72e74a3688e1d2734d28b73b0f48199d87897)

**Comment by dannywillems** [comment]
> Before merging, I would suggest squashing bindings.js changes into only one. In particular, it will remove the ugly `WIP` in the history.

CI seems happy for SnarkyJS. Squashing commits to make the history cleaner.

**Pull request merged**